tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -3391,6 +3391,106 @@ llvm::Optional<Value> convertTFConv2DCommon(
+  // indicate the shape has depth dimension or not.
+  tensorflow::TensorFormat data_format_tf;
+  if (!FormatFromString("NHWC", &data_format_tf)) return llvm::None;
+
+  if(tf_pad == tensorflow::Padding::EXPLICIT) {
+    (void)rewriter.notifyMatchFailure(op, "don't have explicit padding");
+    return llvm::None;
+  }
+
+  ArrayAttr pad;
+  const bool has_depth = true;
+  if (!getPaddingValuesFromPadType(tf_pad, data_format_tf, 0, input_type,
+                                   filter_type, stride, dilation, rewriter,
+                                   pad)) {
+    return llvm::None;
+  }
+
+  // Note that the kernel shape of tfl.conv_3d isn't [O, D, H, W, I] but
+  // [D, H, W, I, O] which is the same as in TF.
+  // Transpose filter shape from [D, H, W, I, O] to [O, D, H, W, C]
+  auto filter_shape = filter_type.getShape();
+  SmallVector<int64_t, 5> a1_transpose_dims;
+  a1_transpose_dims.push_back(filter_shape[4]);
+  a1_transpose_dims.push_back(filter_shape[0]);
+  a1_transpose_dims.push_back(filter_shape[1]);
+  a1_transpose_dims.push_back(filter_shape[2]);
+  a1_transpose_dims.push_back(filter_shape[3]);
+  llvm::Optional<Value> a1_filter_transpose_perm = getConstTensor<int32_t>(
+      rewriter, op, /*vec=*/{4, 0, 1, 2, 3}, /*shape=*/{5});
+
+  if (!a1_filter_transpose_perm) return llvm::None;
+
+  auto a1_filter_transpose_op = CreateOpAndInfer<tosa::TransposeOp>(
+      rewriter, op->getLoc(),
+      RankedTensorType::get(a1_transpose_dims, filter_type.getElementType()),
+      filter, a1_filter_transpose_perm.getValue());
+
+  return CreateOpAndInfer<tosa::Conv3DOp>(
+             rewriter, op->getLoc(), output_type, input,
+             a1_filter_transpose_op.getResult(), bias, pad, stride, dilation)
+      .getResult();
+}
+
+llvm::Optional<Value> convertTFConv3DCommon(
+    PatternRewriter& rewriter, Operation* op, ShapedType output_type,
+    Value input, Value filter, Value bias, ArrayAttr strides_attr, ArrayAttr dilations_attr,
+    StringRef padding_ref, StringRef data_format_ref) {
+
+    ArrayAttr stride;
+    if (!strides_attr) {
+      // Defaults to [1, 1, 1].
+      stride = rewriter.getI64ArrayAttr({1, 1, 1});
+    } else {
+      int64_t stride_d = strides_attr[1].cast<IntegerAttr>().getInt();
+      int64_t stride_h = strides_attr[2].cast<IntegerAttr>().getInt();
+      int64_t stride_w = strides_attr[3].cast<IntegerAttr>().getInt();
+      stride = rewriter.getI64ArrayAttr({stride_d, stride_h, stride_w});
+    }
+
+    ArrayAttr dilation;
+    if (!dilations_attr) {
+      // Defaults to [1, 1, 1].
+      dilation = rewriter.getI64ArrayAttr({1, 1, 1});
+    } else {
+      int64_t dilation_d = dilations_attr[1].cast<IntegerAttr>().getInt();
+      int64_t dilation_h = dilations_attr[2].cast<IntegerAttr>().getInt();
+      int64_t dilation_w = dilations_attr[3].cast<IntegerAttr>().getInt();
+      dilation = rewriter.getI64ArrayAttr({dilation_d, dilation_h, dilation_w});
+    }
+
+   return convertConv3DCommon(
```

Comment:
Did clang-format produce this?

Reply:
Sorry. I didn't run clang-format before this submission. In the next patch, I run clang-format-10 on it.